# Prescriptive Models Assignment 1

## Setup
First import the necessary python modules and libraries.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import statsmodels.api as sm
import statsmodels.formula.api as smf

## 1. Table 1

### Question 1:
*Load the “charitable giving.csv” dataset and run a regression to assess whether the average
“Number of months since last donation” is significantly different between treatment and control. Interpret
the relevant regression coefficients and compare the regression-based comparison to the group-specific means
reported in Table 1 of the paper.*

In [13]:
# Load the charitable giving dataset
df = pd.read_csv('charitable_giving.csv')
df.head()

donation_amount  donation_dummy  control  treatment  match_ratio  ratio1  \
0              0.0             0.0      0.0        1.0          1.0       1   
1              0.0             0.0      1.0        0.0          0.0       0   
2              0.0             0.0      1.0        0.0          0.0       0   
3              0.0             0.0      0.0        1.0          3.0       0   
4              0.0             0.0      0.0        1.0          2.0       0   

   ratio2  ratio3  red_state_dummy  months_since_last_donation  \
0     0.0     0.0              1.0                        19.0   
1     0.0     0.0              1.0                        29.0   
2     0.0     0.0              1.0                         3.0   
3     0.0     1.0              0.0                         4.0   
4     1.0     0.0              0.0                         8.0   

   highest_previous_donation  prior_donations_num  
0                      500.0                 32.0  
1                      300.0                 22.0  
2                      500.0                 22.0  
3                      250.0                 29.0  
4                       50.0                 17.0

Below is the regression of average "Number of months since last donation" against the treatment group. Variable is equal to 1 if it's treatment, 0 if it's control.

In [4]:
model_months = smf.ols(formula = 'months_since_last_donation ~ treatment',data = df)
res_months = model_months.fit()
print(res_months.summary())

                                OLS Regression Results                                
Dep. Variable:     months_since_last_donation   R-squared:                       0.000
Model:                                    OLS   Adj. R-squared:                 -0.000
Method:                         Least Squares   F-statistic:                   0.01428
Date:                        Mon, 17 Jan 2022   Prob (F-statistic):              0.905
Time:                                19:58:30   Log-Likelihood:            -1.9585e+05
No. Observations:                       50082   AIC:                         3.917e+05
Df Residuals:                           50080   BIC:                         3.917e+05
Df Model:                                   1                                         
Covariance Type:                    nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------

Average number of months since last donation for the treatment group is equal to $12.9981 + 0.0137 = 13.0118$ since we set the variable for treatment equal to 1 and include the $0.0137$ coefficient value in our answer. On the other hand, the average number of months since last donation for the control group is equal to just $12.9981$ since we do not include in the $0.0137$ in our answer because it's the control group.

The answers we achieve from our regression compare well to the group-specific means reported in Table 1 which have values of $13.012$ for treatment and $12.998$ for control.

### Question 2: 
*Is the difference in “Number of month since last donation” between treatment and control statistically significant (at the usual 95% confidence level)? Is this the result you expected?*


At the 95% confidence level, the difference between treatment and control is not statistically significant as we can see our p-value of $0.905$ does not fall below the necessary $0.05$ level to be considered significant.

This is the result I expected because when you are testing treatment groups against control groups, the two groups should be very similar in their observable demographic information and prior history so that the effect of the treatment can be properly measured. Therefore, there should be no significant difference between the two groups on this statistic.

### Question 3:
*More generally, describe the take-away from Table 1 in the paper.*

Table 1 is essentially just to provide us with some summary statistics of our treatment and control groups to demonstrate the reliability of the sampling methods. In addition, any statistical significance found between treatment and control groups as a result of the experiment could be boiled down to something other than historical or demographic differences between the treatment and control groups.

## 2. Response rate regressions

### Question 1:
*Run a linear regression of response rate (the donation dummy) on the treatment dummy (and
an intercept). Interpret both coefficients and compare them to the results presented in the first row of Table 2a.*

In [8]:
model_donation = smf.ols(formula = 'donation_dummy ~ treatment',data = df)
res_donation = model_donation.fit()
print(res_donation.summary())

                            OLS Regression Results                            
Dep. Variable:         donation_dummy   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     9.618
Date:                Mon, 17 Jan 2022   Prob (F-statistic):            0.00193
Time:                        20:52:24   Log-Likelihood:                 26630.
No. Observations:               50083   AIC:                        -5.326e+04
Df Residuals:                   50081   BIC:                        -5.324e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0179      0.001     16.225      0.0

Response rate for the treatment group is equal to $0.0179 + 0.0042 = 0.0221$ since we set the variable for treatment equal to 1 and include the $0.0042$ coefficient value in our answer. On the other hand, the response rate for the control group is equal to just $0.0179$ since we do not include in the $0.0042$ in our answer because it's the control group.

The answers we achieve from our regression compare well to the results reported in Table 2A which have values of $0.022$ for treatment and $0.018$ for control.

### Question 2:
*Run a regression on three dummies for match ratio treatment (1:1, 2:1, and 3:1 and an
intercept). Interpret all four regression coefficients.*

In [9]:
model_ratio = smf.ols(formula = 'donation_dummy ~ ratio1 + ratio2 + ratio3',data = df)
res_ratio = model_ratio.fit()
print(res_ratio.summary())

                            OLS Regression Results                            
Dep. Variable:         donation_dummy   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     3.665
Date:                Mon, 17 Jan 2022   Prob (F-statistic):             0.0118
Time:                        21:01:53   Log-Likelihood:                 26630.
No. Observations:               50083   AIC:                        -5.325e+04
Df Residuals:                   50079   BIC:                        -5.322e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0179      0.001     16.225      0.0

Our intercept coefficient of $0.0179$ stays the same as before and indicates response rate for control group only. Coefficients for each of the match ratios are positive which indicates that there's an increase in response rate for all treatment groups, regardless of the level of the match ratio. In addition, if we'd like to assess the response rate for an individual match ratio, we would assign one of those variables a value of 1 and assign the rest of the variables a value of 0.

### Question 3:
*Calculate the response rate difference between the 1:1 and 2:1 match ratios.*

Response rate for the 1:1 match ratio is equal to $0.0179 + 0.0029 = 0.0208$ while the response rate for the 2:1 match ratio is equal to $0.0179 + 0.0048 = 0.0227$. Taking the difference between these two, we get $0.0227 - 0.0208 = 0.0019$. Thus, there is a 0.0019 difference in response rate between the 1:1 and 2:1 match ratios.

### Question 4:
*Based on the regressions you just ran and more generally the results in Table 2a, what do you
conclude regarding the effectiveness of using matched donations?*


Our regressions and the figures in Table 2A suggest that while there is possibly a significant difference in response rate between the treatment and control groups (at the 90% significance level), there does not seem to be any meaningful difference in donation likelihood among the match ratios at least. Meaning, larger match ratios relative to smaller match ratios have no additional influence.

## 3. Response rates in red/blue states

### Question 1:
*Repeat the regression of response rate on treatment and an intercept (do not include separate
match ratio dummies). But this time, base the regression only on respondents in blue states or red states.
I.e. run two regressions, one on each of the two sub-samples of data. Interpret the coefficients in both
regressions. Is the treatment more effective in red or blue states?*




**Regression on blue states:**

In [12]:
# Create sub-sample of dataframe for blue states
df_blue = df[df["red_state_dummy"] == 0]
df_blue.head()

donation_amount  donation_dummy  control  treatment  match_ratio  ratio1  \
3               0.0             0.0      0.0        1.0          3.0       0   
4               0.0             0.0      0.0        1.0          2.0       0   
5               0.0             0.0      0.0        1.0          1.0       1   
6               0.0             0.0      1.0        0.0          0.0       0   
12              0.0             0.0      0.0        1.0          3.0       0   

    ratio2  ratio3  red_state_dummy  months_since_last_donation  \
3      0.0     1.0              0.0                         4.0   
4      1.0     0.0              0.0                         8.0   
5      0.0     0.0              0.0                        26.0   
6      0.0     0.0              0.0                        28.0   
12     0.0     1.0              0.0                        30.0   

    highest_previous_donation  prior_donations_num  
3                       250.0                 29.0  
4                        50.0                 17.0  
5                        25.0                  4.0  
6                        35.0                 10.0  
12                       35.0                 14.0

In [15]:
model_blue = smf.ols(formula = 'donation_dummy ~ treatment',data = df_blue)
res_blue = model_blue.fit()
print(res_blue.summary())

                            OLS Regression Results                            
Dep. Variable:         donation_dummy   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.3567
Date:                Mon, 17 Jan 2022   Prob (F-statistic):              0.550
Time:                        21:35:31   Log-Likelihood:                 15783.
No. Observations:               29806   AIC:                        -3.156e+04
Df Residuals:                   29804   BIC:                        -3.155e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0200      0.001     14.085      0.0

**Regression on red states:**

In [14]:
# Create sub-sample of dataframe for red states
df_red = df[df["red_state_dummy"] == 1]
df_red.head()

donation_amount  donation_dummy  control  treatment  match_ratio  ratio1  \
0              0.0             0.0      0.0        1.0          1.0       1   
1              0.0             0.0      1.0        0.0          0.0       0   
2              0.0             0.0      1.0        0.0          0.0       0   
7              0.0             0.0      0.0        1.0          2.0       0   
8              0.0             0.0      0.0        1.0          3.0       0   

   ratio2  ratio3  red_state_dummy  months_since_last_donation  \
0     0.0     0.0              1.0                        19.0   
1     0.0     0.0              1.0                        29.0   
2     0.0     0.0              1.0                         3.0   
7     1.0     0.0              1.0                        11.0   
8     0.0     1.0              1.0                        30.0   

   highest_previous_donation  prior_donations_num  
0                      500.0                 32.0  
1                      300.0                 22.0  
2                      500.0                 22.0  
7                      300.0                 20.0  
8                      100.0                  8.0

In [16]:
model_red = smf.ols(formula = 'donation_dummy ~ treatment',data = df_red)
res_red = model_red.fit()
print(res_red.summary())

                            OLS Regression Results                            
Dep. Variable:         donation_dummy   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                  0.001
Method:                 Least Squares   F-statistic:                     17.24
Date:                Mon, 17 Jan 2022   Prob (F-statistic):           3.31e-05
Time:                        21:35:35   Log-Likelihood:                 10839.
No. Observations:               20242   AIC:                        -2.167e+04
Df Residuals:                   20240   BIC:                        -2.166e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.0146      0.002      8.398      0.0

The coefficients in both our regressions for the blue and red states are all positive which indicates there's an increase in response rate for the treatment groups. Specifically, the blue states have a value of $0.0200$ for the control group and a value of $0.0210$ for the treatment group, while the red states have a value of $0.0146$ for the control group and a value of $0.0234$ for the treatment group.

Based on our regressions, we can see that the treatment is more effective in the red states as the coefficient on the treatment variable is statistically significant at the 95% confidence level (and above) while the coefficient on the treatment variable for the blue states is not statistically significant.

### Question 2:
*States are of course not randomly assigned. Does the treatment coefficient have a causal
interpretation in each of the two regressions? Does the difference in the treatment effect between states
have a causal interpretation?*

When looking at each of the regressions separately, we can say that the treatment coefficient has a causal interpretation within those blue states alone or red states alone (treatment coefficient is significant/insignificant). Because the states are not randomly assigned however, the difference in the treatment effect does not have a causal interpretation because the demographics of each group could be different and there could be other factors influencing the difference in results that have nothing to do with the treatment (a.k.a. omitted variable bias).

## 4. Response rates and donation amount

### Question 1: 
*Run a regression of dollars given on a treatment dummy and an intercept. Interpret the
regression coefficients. Does the treatment coefficient have a causal interpretation?*



In [17]:
model_dollars = smf.ols(formula = 'donation_amount ~ treatment',data = df)
res_dollars = model_dollars.fit()
print(res_dollars.summary())

                            OLS Regression Results                            
Dep. Variable:        donation_amount   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     3.461
Date:                Mon, 17 Jan 2022   Prob (F-statistic):             0.0628
Time:                        22:10:06   Log-Likelihood:            -1.7946e+05
No. Observations:               50083   AIC:                         3.589e+05
Df Residuals:                   50081   BIC:                         3.589e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.8133      0.067     12.063      0.0

The intercept regression coefficient represents the control group while the treatment variable indicates the effect on donation amount for those in the treatment group. On average, we can expect a value of $0.8133$ for the donation amount of those in the control group and a value of $0.8133 + 0.1536 = 0.9669$ for the donation amount of those in the treatment group. Positive value for the treatment coefficient indicates that the treatment effect increases donation amount on average.

In this situation, the treatment coefficient does have a causal interpretation because the demographics between control and treatment groups are controlled to be similar to one another. Therefore, we can assess any significant difference between the two groups to be a result of the treatment effect as opposed to observational differences between the two group's demographics.

### Question 2:
*Next, regress dollars given on a treatment dummy and an intercept, but base the regression
only on respondents that made a donation (i.e. donation dummy is equal to 1). This regression allows
you to analyze how much respondents donate conditional on donating some positive amount. Interpret the
regression coefficients. Does the treatment coefficient have a causal interpretation?*

In [21]:
# Create sub-sample of dataframe for those that made a donation
df_donate = df[df["donation_dummy"] == 1]
df_donate.head()

donation_amount  donation_dummy  control  treatment  match_ratio  ratio1  \
24             40.0             1.0      0.0        1.0          2.0       0   
31             25.0             1.0      0.0        1.0          1.0       1   
51             25.0             1.0      0.0        1.0          1.0       1   
83             25.0             1.0      0.0        1.0          1.0       1   
88             15.0             1.0      0.0        1.0          3.0       0   

    ratio2  ratio3  red_state_dummy  months_since_last_donation  \
24     1.0     0.0              1.0                         7.0   
31     0.0     0.0              0.0                         8.0   
51     0.0     0.0              0.0                        30.0   
83     0.0     0.0              0.0                         2.0   
88     0.0     1.0              0.0                        10.0   

    highest_previous_donation  prior_donations_num  
24                      100.0                 41.0  
31                       25.0                 19.0  
51                      200.0                 14.0  
83                       50.0                 40.0  
88                       50.0                 39.0

In [19]:
model_donation2 = smf.ols(formula = 'donation_amount ~ treatment',data = df_donate)
res_donation2 = model_donation2.fit()
print(res_donation2.summary())

                            OLS Regression Results                            
Dep. Variable:        donation_amount   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.3374
Date:                Mon, 17 Jan 2022   Prob (F-statistic):              0.561
Time:                        23:27:55   Log-Likelihood:                -5326.8
No. Observations:                1034   AIC:                         1.066e+04
Df Residuals:                    1032   BIC:                         1.067e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     45.5403      2.423     18.792      0.0

The intercept regression coefficient represents the control group while the treatment variable indicates the effect on donation amount for those in the treatment group. On average, we can expect a value of  $45.5403$  for the donation amount of those in the control group and a value of  $45.5403-1.6684=43.8719$  for the donation amount of those in the treatment group. Negative value for the treatment coefficient indicates that the treatment effect decreases donation amount on average among those who made any donation.

In this situation, the treatment coefficient does not have a causal interpretation because the demographics between control and treatment groups are not controlled to be similar when we take a subset of the original data. Therefore, we cannot assess any significant difference between the two groups to be a result of the treatment effect as opposed to observational differences between the two group's demographics.